cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java

```diff
@@ -318,47 +482,70 @@ boolean tryWrite() {
       // The visibility of states mutation done while getting hold of the writerFlag,
       // is piggy back on the writerFlag atomic variable update in the finally block,
       // hence all states mutated will be visible to all threads after that.
-      int bytesWritten = 0;
-      int eventsWritten = 0;
-      List<HandlerStreamEventData> processQueue = Lists.newArrayListWithExpectedSize(workerThreads);
       try {
-        FileWriter<StreamEvent> writer = writerSupplier.get();
-        HandlerStreamEventData data = queue.poll();
-        long timestamp = System.currentTimeMillis();
-        while (data != null) {
-          processQueue.add(data);
-          writer.append(streamEvent.set(data, timestamp));
-          data = queue.poll();
-        }
-        writer.flush();
-        for (HandlerStreamEventData processed : processQueue) {
-          processed.completed(null);
-          bytesWritten += processed.getBody().remaining();
-        }
-        eventsWritten = processQueue.size();
-      } catch (Throwable t) {
-        LOG.error("Failed to write to file for stream {}.", streamName, t);
-        // On exception, remove this EventQueue from the map and close the writer associated with this instance
-        eventQueues.remove(streamName, this);
-        Closeables.closeQuietly(writerSupplier.get());
-
-        for (HandlerStreamEventData processed : processQueue) {
-          processed.completed(t);
+        metrics.reset();
+        List<WriteRequest> processQueue = Lists.newArrayListWithExpectedSize(workerThreads);
+        try {
+          FileWriter<StreamEventData> writer = getFileWriter();
+          WriteRequest request = queue.poll();
+          streamEvent.setTimestamp(System.currentTimeMillis());
+          while (request != null) {
+            processQueue.add(request);
+            request.write(writer, metrics);
+            request = queue.poll();
+          }
+          writer.flush();
+          for (WriteRequest processed : processQueue) {
+            processed.completed(null);
+          }
+          bytesWritten = metrics.bytesWritten;
+          eventsWritten = metrics.eventsWritten;
+        } catch (Throwable t) {
+          // On exception, remove this EventQueue from the map and close this event queue
+          eventQueues.remove(streamName, this);
+          doClose();
+
+          for (WriteRequest processed : processQueue) {
+            processed.completed(t);
+          }
         }
       } finally {
         writerFlag.set(false);
       }
 
+      emitMetrics(bytesWritten, eventsWritten);
+      return true;
+    }
+
+    private void emitMetrics(long bytesWritten, long eventsWritten) {
+      // TODO: Blocked by CDAP-1189 to take long value for increment
+      if (bytesWritten > 0) {
+        metricsCollector.increment("collect.bytes", (int) bytesWritten);
```

Comment:
what happens if we send a negative number due to overflow? unless CDAP-1189 will be fixed for the release, seems like we should either not emit or break it into multiple increments.